src/charm.py

```diff
@@ -0,0 +1,257 @@
+        container.replan()
+
+    def _on_storage_attached(self, event: ops.StorageAttachedEvent) -> None:
+        self._push_library_to_storage()
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Don't do anything! User can run library-write action if needed.
+
+        Bad config values are handled in _on_collect_status
+        """
+        pass
+
+    def _on_library_write(self, event: ops.ActionEvent) -> None:
+        library_write_behaviour, status = self._get_library_write_behaviour()
+        if library_write_behaviour is None:
+            event.fail(status.message)
+            return
+        self._push_library_to_storage()
+        event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})
+
+    def _on_library_info(self, event: ops.ActionEvent) -> None:
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
+        format = cast(str, event.params[LIBRARY_INFO_FORMAT_PARAM])
+        match format:
+            case 'tree':
+                try:
+                    container.exec(['which', 'tree']).wait()
+                except ops.pebble.ExecError:
+                    logger.debug('_on_library_info: tree: installing dependencies')
+                    container.exec(['apt', 'update']).wait()
+                    container.exec(['apt', 'install', 'tree', '-y']).wait()
+                logger.debug('_on_library_info: tree: executing')
+                process = container.exec(
+                    ['tree'],
+                    working_dir='/books/',
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'tree': '\n'.join(stdout.lines)})
+            case 'ls-1':
+                logger.debug('_on_library_info: tree: executing')
+                process = container.exec(
+                    ['ls', '-1'],
+                    working_dir='/books/',
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'ls-1': '\n'.join(stdout.lines)})
+            case _:
+                msg = (
+                    f'Invalid value {format} for {LIBRARY_INFO_FORMAT_PARAM} parameter'
+                    f' of {LIBRARY_INFO_ACTION} action.'
+                )
+                logger.error(f'_on_library_info: {format}: {msg}')
+                event.fail(msg)
+
+    @staticmethod
+    def get_pebble_layer() -> CalibreWebLayerDict:
+        """Return a dictionary representing the Pebble layer."""
+        c = ' && '.join(
+            [
+                'bash /etc/s6-overlay/s6-rc.d/init-calibre-web-config/run',
+                # with bash because the run script shebang depends on s6
+                'export CALIBRE_DBPATH=/config',
+                'cd /app/calibre-web',
+                'python3 /app/calibre-web/cps.py',
```

Comment:
The `bash .../run` line runs a script that would normally be run by `s6-overlay` when deploying this docker image. It creates some files and directories and sets some permissions. Now that you mention it, I guess this charm only targets this version of the docker image, since there are a whole bunch of other places the authors could add various important things that `s6-overlay` would be expected to run. (It refuses to run in this case because it's not pid 1 (pebble is)). I think in the long term I'd ideally just create a new Dockerfile for the charm without any of the `s6` stuff.

Tony also pointed out that env and working-dir can be handled in the pebble layer, I'll try that.